pkg/runtime/wfengine/workflow.go

```diff
@@ -0,0 +1,251 @@
+	scheduler    workflowScheduler
+}
+
+type workflowState struct {
+	inbox   []*backend.HistoryEvent
+	history []*backend.HistoryEvent
+
+	// runtimeState is a cached copy of the [backend.OrchestrationRuntimeState]
+	runtimeState *backend.OrchestrationRuntimeState
+}
+
+const (
+	CallbackChannelProperty = "dapr.callback"
+
+	CreateWorkflowInstanceMethod = "CreateWorkflowInstance"
+	GetWorkflowMetadataMethod    = "GetWorkflowMetadata"
+
+	createWorkflowReminder = "start" // internal reminder for executing workflows
+)
+
+func NewWorkflowActor(scheduler workflowScheduler) actors.InternalActor {
+	return &workflowActor{
+		states:    make(map[string]*workflowState),
+		scheduler: scheduler,
+	}
+}
+
+// SetActorRuntime implements actors.InternalActor
+func (wf *workflowActor) SetActorRuntime(actorRuntime actors.Actors) {
+	wf.actorRuntime = actorRuntime
+}
+
+// InvokeMethod implements actors.InternalActor
+func (wf *workflowActor) InvokeMethod(ctx context.Context, actorID string, methodName string, request []byte) (interface{}, error) {
+	wfLogger.Debugf("invoking method '%s' on workflow actor '%s'", methodName, actorID)
+
+	var result interface{}
+	var err error
+
+	switch methodName {
+	case CreateWorkflowInstanceMethod:
+		err = wf.createWorkflowInstanceMethod(ctx, actorID, request)
+	case GetWorkflowMetadataMethod:
+		result, err = wf.getWorkflowMetadata(actorID)
+	default:
+		err = fmt.Errorf("no such method: %s", methodName)
+	}
+	return result, err
+}
+
+// InvokeReminder implements actors.InternalActor
+func (wf *workflowActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, params []byte) error {
+	wfLogger.Debugf("invoking reminder '%s' on workflow actor '%s'", reminderName, actorID)
+
+	var err error
+	switch reminderName {
+	case createWorkflowReminder:
+		// Workflow executions should never take longer than a few seconds at the most
+		timeoutCtx, cancelTimeout := context.WithTimeout(ctx, 30*time.Second) // TODO: Configurable
+		defer cancelTimeout()
+		err = wf.runWorkflow(timeoutCtx, actorID)
+	default:
+		wfLogger.Warnf("reminder '%s' for workflow actor '%s' was not recognized", reminderName, actorID)
+	}
+
+	return err
+}
+
+// InvokeTimer implements actors.InternalActor
+func (wf *workflowActor) InvokeTimer(ctx context.Context, actorID string, timerName string, params []byte) error {
+	wfLogger.Debugf("invoking timer '%s' on workflow actor '%s'", timerName, actorID)
+	// TODO
+	return nil
+}
+
+// DeactivateActor implements actors.InternalActor
+func (wf *workflowActor) DeactivateActor(ctx context.Context, actorID string) error {
+	wfLogger.Debugf("deactivating workflow actor '%s'", actorID)
+	delete(wf.states, actorID)
+	return nil
+}
+
+func (wf *workflowActor) createWorkflowInstanceMethod(ctx context.Context, actorID string, startEventBytes []byte) error {
+	// create a new state entry if one doesn't already exist
+	state, exists, err := wf.loadInternalState(actorID)
```

Comment:
loadInternalState only ever returns nil for the err. Is the check on "if err != nil" below to account for the updating of the loadInternalState function in the future? I saw that there's a todo line on the return in that function:
`return nil, false, nil // TODO: Implement fetching from storage after validating saving to storage`
Is that "TODO" line going to cover the possibility of a non-nil err return?

Reply:
Yes, exactly. I haven't yet implemented the fetching and decoding of fetched state, and those are opportunities for us to return non-nil error values.